### PR TITLE
fix: Unable to deprecate bundles that are head of a NOT default channel with opm deprecatetruncate

### PR DESCRIPTION
**Pull request by camilamacedo86**
**Description**
Fix Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=2001940

**What was the problem**
The wrong syntax with inner join returns the values for ANY head of channel and not only when it is the head of the channel which is the default of the package.

**Problem faced (Example scenario)**

Unable to deprecate the following bundles from the image registry.redhat.io/redhat/community-operator-index:v4.8. 

apicurio-registry.v0.0.1 - defaultChannel:NO
apicurio-registry.v0.0.3-v1.2.3.final - defaultChannel:NO
apicurio-registry.v0.0.4-v1.3.2.final - defaultChannel:NO

When should deprecate since the image has the bundle: apicurio-registry-operator.v1.0.0-v2.0.0.final which is the head of the default channel 2.x which was not deprecated.

**Testing the changes made in this PR**

1) Runing the command : 

```
./bin/opm index deprecatetruncate --bundles="quay.io/openshift-community-operators/apicurio-registry@sha256:661b4538ba60b7e25754d82132590a9e10182f38679db1011abe81c161e8903e,quay.io/openshift-community-operators/apicurio-registry@sha256:85f51fd88be01507d8f8ae9032066dee232960715a8b6e97f2b4c6d5d3cc9671,quay.io/openshift-community-operators/apicurio-registry@sha256:239525b1a4a7030fb0184bea77a1e56b9bb652bbbb78e3facdedf829ee2f7e9d" --from-index=registry.redhat.io/redhat/community-operator-index:v4.8  --tag "quay/cmacedo/community-operator-index:apicu" --allow-package-removal --container-tool=docker
Password:
Sorry, try again.
Password:
WARN[0000] DEPRECATION NOTICE:
Sqlite-based catalogs and their related subcommands are deprecated. Support for
them will be removed in a future release. Please migrate your catalog workflows
to the new file-based catalog format. 
INFO[0000] deprecating bundles from the index            bundles="[quay.io/openshift-community-operators/apicurio-registry@sha256:661b4538ba60b7e25754d82132590a9e10182f38679db1011abe81c161e8903e quay.io/openshift-community-operators/apicurio-registry@sha256:85f51fd88be01507d8f8ae9032066dee232960715a8b6e97f2b4c6d5d3cc9671 quay.io/openshift-community-operators/apicurio-registry@sha256:239525b1a4a7030fb0184bea77a1e56b9bb652bbbb78e3facdedf829ee2f7e9d]"
INFO[0000] Pulling previous image registry.redhat.io/redhat/community-operator-index:v4.8 to get metadata  bundles="[quay.io/openshift-community-operators/apicurio-registry@sha256:661b4538ba60b7e25754d82132590a9e10182f38679db1011abe81c161e8903e quay.io/openshift-community-operators/apicurio-registry@sha256:85f51fd88be01507d8f8ae9032066dee232960715a8b6e97f2b4c6d5d3cc9671 quay.io/openshift-community-operators/apicurio-registry@sha256:239525b1a4a7030fb0184bea77a1e56b9bb652bbbb78e3facdedf829ee2f7e9d]"
INFO[0000] running /usr/local/bin/docker pull registry.redhat.io/redhat/community-operator-index:v4.8  bundles="[quay.io/openshift-community-operators/apicurio-registry@sha256:661b4538ba60b7e25754d82132590a9e10182f38679db1011abe81c161e8903e quay.io/openshift-community-operators/apicurio-registry@sha256:85f51fd88be01507d8f8ae9032066dee232960715a8b6e97f2b4c6d5d3cc9671 quay.io/openshift-community-operators/apicurio-registry@sha256:239525b1a4a7030fb0184bea77a1e56b9bb652bbbb78e3facdedf829ee2f7e9d]"
INFO[0004] running /usr/local/bin/docker pull registry.redhat.io/redhat/community-operator-index:v4.8  bundles="[quay.io/openshift-community-operators/apicurio-registry@sha256:661b4538ba60b7e25754d82132590a9e10182f38679db1011abe81c161e8903e quay.io/openshift-community-operators/apicurio-registry@sha256:85f51fd88be01507d8f8ae9032066dee232960715a8b6e97f2b4c6d5d3cc9671 quay.io/openshift-community-operators/apicurio-registry@sha256:239525b1a4a7030fb0184bea77a1e56b9bb652bbbb78e3facdedf829ee2f7e9d]"
INFO[0006] Getting label data from previous image        bundles="[quay.io/openshift-community-operators/apicurio-registry@sha256:661b4538ba60b7e25754d82132590a9e10182f38679db1011abe81c161e8903e quay.io/openshift-community-operators/apicurio-registry@sha256:85f51fd88be01507d8f8ae9032066dee232960715a8b6e97f2b4c6d5d3cc9671 quay.io/openshift-community-operators/apicurio-registry@sha256:239525b1a4a7030fb0184bea77a1e56b9bb652bbbb78e3facdedf829ee2f7e9d]"
INFO[0006] running docker inspect                        bundles="[quay.io/openshift-community-operators/apicurio-registry@sha256:661b4538ba60b7e25754d82132590a9e10182f38679db1011abe81c161e8903e quay.io/openshift-community-operators/apicurio-registry@sha256:85f51fd88be01507d8f8ae9032066dee232960715a8b6e97f2b4c6d5d3cc9671 quay.io/openshift-community-operators/apicurio-registry@sha256:239525b1a4a7030fb0184bea77a1e56b9bb652bbbb78e3facdedf829ee2f7e9d]"
INFO[0006] running docker create                         bundles="[quay.io/openshift-community-operators/apicurio-registry@sha256:661b4538ba60b7e25754d82132590a9e10182f38679db1011abe81c161e8903e quay.io/openshift-community-operators/apicurio-registry@sha256:85f51fd88be01507d8f8ae9032066dee232960715a8b6e97f2b4c6d5d3cc9671 quay.io/openshift-community-operators/apicurio-registry@sha256:239525b1a4a7030fb0184bea77a1e56b9bb652bbbb78e3facdedf829ee2f7e9d]"
INFO[0006] running docker cp                             bundles="[quay.io/openshift-community-operators/apicurio-registry@sha256:661b4538ba60b7e25754d82132590a9e10182f38679db1011abe81c161e8903e quay.io/openshift-community-operators/apicurio-registry@sha256:85f51fd88be01507d8f8ae9032066dee232960715a8b6e97f2b4c6d5d3cc9671 quay.io/openshift-community-operators/apicurio-registry@sha256:239525b1a4a7030fb0184bea77a1e56b9bb652bbbb78e3facdedf829ee2f7e9d]"
INFO[0017] running docker rm                             bundles="[quay.io/openshift-community-operators/apicurio-registry@sha256:661b4538ba60b7e25754d82132590a9e10182f38679db1011abe81c161e8903e quay.io/openshift-community-operators/apicurio-registry@sha256:85f51fd88be01507d8f8ae9032066dee232960715a8b6e97f2b4c6d5d3cc9671 quay.io/openshift-community-operators/apicurio-registry@sha256:239525b1a4a7030fb0184bea77a1e56b9bb652bbbb78e3facdedf829ee2f7e9d]"
INFO[0019] allow-package-removal enabled: checking default channel heads for package removal  bundles="[quay.io/openshift-community-operators/apicurio-registry@sha256:661b4538ba60b7e25754d82132590a9e10182f38679db1011abe81c161e8903e quay.io/openshift-community-operators/apicurio-registry@sha256:85f51fd88be01507d8f8ae9032066dee232960715a8b6e97f2b4c6d5d3cc9671 quay.io/openshift-community-operators/apicurio-registry@sha256:239525b1a4a7030fb0184bea77a1e56b9bb652bbbb78e3facdedf829ee2f7e9d]"
INFO[0019] no head of default channel found - skipping package removal  bundles="[quay.io/openshift-community-operators/apicurio-registry@sha256:661b4538ba60b7e25754d82132590a9e10182f38679db1011abe81c161e8903e quay.io/openshift-community-operators/apicurio-registry@sha256:85f51fd88be01507d8f8ae9032066dee232960715a8b6e97f2b4c6d5d3cc9671 quay.io/openshift-community-operators/apicurio-registry@sha256:239525b1a4a7030fb0184bea77a1e56b9bb652bbbb78e3facdedf829ee2f7e9d]"
INFO[0019] deprecating bundles                           bundles="[quay.io/openshift-community-operators/apicurio-registry@sha256:661b4538ba60b7e25754d82132590a9e10182f38679db1011abe81c161e8903e quay.io/openshift-community-operators/apicurio-registry@sha256:85f51fd88be01507d8f8ae9032066dee232960715a8b6e97f2b4c6d5d3cc9671 quay.io/openshift-community-operators/apicurio-registry@sha256:239525b1a4a7030fb0184bea77a1e56b9bb652bbbb78e3facdedf829ee2f7e9d]"
INFO[0019] Generating dockerfile                         bundles="[quay.io/openshift-community-operators/apicurio-registry@sha256:661b4538ba60b7e25754d82132590a9e10182f38679db1011abe81c161e8903e quay.io/openshift-community-operators/apicurio-registry@sha256:85f51fd88be01507d8f8ae9032066dee232960715a8b6e97f2b4c6d5d3cc9671 quay.io/openshift-community-operators/apicurio-registry@sha256:239525b1a4a7030fb0184bea77a1e56b9bb652bbbb78e3facdedf829ee2f7e9d]"
INFO[0019] writing dockerfile: index.Dockerfile351495467  bundles="[quay.io/openshift-community-operators/apicurio-registry@sha256:661b4538ba60b7e25754d82132590a9e10182f38679db1011abe81c161e8903e quay.io/openshift-community-operators/apicurio-registry@sha256:85f51fd88be01507d8f8ae9032066dee232960715a8b6e97f2b4c6d5d3cc9671 quay.io/openshift-community-operators/apicurio-registry@sha256:239525b1a4a7030fb0184bea77a1e56b9bb652bbbb78e3facdedf829ee2f7e9d]"
INFO[0019] running docker build                          bundles="[quay.io/openshift-community-operators/apicurio-registry@sha256:661b4538ba60b7e25754d82132590a9e10182f38679db1011abe81c161e8903e quay.io/openshift-community-operators/apicurio-registry@sha256:85f51fd88be01507d8f8ae9032066dee232960715a8b6e97f2b4c6d5d3cc9671 quay.io/openshift-community-operators/apicurio-registry@sha256:239525b1a4a7030fb0184bea77a1e56b9bb652bbbb78e3facdedf829ee2f7e9d]"
INFO[0019] [docker build -f index.Dockerfile351495467 -t quay/cmacedo/community-operator-index:apicu .]  bundles="[quay.io/openshift-community-operators/apicurio-registry@sha256:661b4538ba60b7e25754d82132590a9e10182f38679db1011abe81c161e8903e quay.io/openshift-community-operators/apicurio-registry@sha256:85f51fd88be01507d8f8ae9032066dee232960715a8b6e97f2b4c6d5d3cc9671 quay.io/openshift-community-operators/apicurio-registry@sha256:239525b1a4a7030fb0184bea77a1e56b9bb652bbbb78e3facdedf829ee2f7e9d]"

```
2) Checking if all bundles were deprecated, the alpha channel was removed and the head of the default package channel was not removed as its channel: 

```
docker create --name test quay/cmacedo/community-operator-index:apicu "yes"
docker cp test:/database/index.db .
open index.db
```

Screen Shot 2021-09-16 at 15 59 34
Screen Shot 2021-09-16 at 15 59 16
Screen Shot 2021-09-16 at 15 57 44

**Regards usage of the method changed in this pr and possible impact which would require regression tests:**

- The places changed in this PR are only used in the specific context of deprecating bundle and I could not find any usage  of them in any other operation
- They are covered with tests and the checked scenario was added too. 

Screen Shot 2021-09-16 at 16 47 40
 
Screen Shot 2021-09-16 at 16 48 30

### CODE REVIEW

**Comment by camilamacedo86** [comment]
c/c @gallettilance @kevinrizza @ankitathomas @njhale

File: pkg/sqlite/query.go [hunk 1]
```diff
@@ -1342,9 +1342,10 @@ func (s *SQLQuerier) listBundleChannels(ctx context.Context, bundleName string)
 // PackageFromDefaultChannelHeadBundle returns the package name if the provided bundle is the head of its default channel.
 func (s *SQLQuerier) PackageFromDefaultChannelHeadBundle(ctx context.Context, bundle string) (string, error) {
 	packageFromDefaultChannelHeadBundle := `
-	SELECT package_name FROM package 
-	INNER JOIN channel ON channel.name = package.default_channel
-	WHERE channel.head_operatorbundle_name = (SELECT name FROM operatorbundle WHERE bundlepath=? LIMIT 1) `
+	SELECT package_name FROM package, channel 
+	WHERE channel.package_name == package.name
+	AND package.default_channel == channel.name
+	AND channel.head_operatorbundle_name = (SELECT name FROM operatorbundle WHERE bundlepath=? LIMIT 1) `
```
Comment (camilamacedo86) [inline pkg/sqlite/query.go:1348]: See that it was returning the package when the bundle is HEAD of channel but the channel is NOT the default channel of the package, e.g:

```
SELECT package_name FROM package 
	INNER JOIN channel ON channel.name = package.default_channel
	WHERE channel.head_operatorbundle_name = (SELECT name FROM operatorbundle WHERE bundlepath='quay.io/openshift-community-operators/apicurio-registry@sha256:239525b1a4a7030fb0184bea77a1e56b9bb652bbbb78e3facdedf829ee2f7e9d' LIMIT 1)
```

Screen Shot 2021-09-16 at 16 32 57

The select with inner join was wrong.

File: pkg/sqlite/load.go [hunk 2]
```diff
@@ -1320,9 +1320,10 @@ func (s *sqlLoader) rmChannelEntry(tx *sql.Tx, csvName string) error {
 
 func getTailFromBundle(tx *sql.Tx, head string) (bundles []string, err error) {
 	getReplacesSkips := `SELECT replaces, skips FROM operatorbundle WHERE name=?`
-	isDefaultChannelHead := `SELECT head_operatorbundle_name FROM channel
-							INNER JOIN package ON channel.name = package.default_channel
-							WHERE channel.head_operatorbundle_name = ?`
+	isDefaultChannelHead := `SELECT head_operatorbundle_name FROM package, channel
+								WHERE channel.package_name == package.name
+								AND package.default_channel == channel.name
+								AND channel.head_operatorbundle_name = ?`
```
Comment (camilamacedo86) [inline pkg/sqlite/load.go:1326]: see that the select with the inner join is wrong and is returning the head_operatorbundle_name when the bundle is head of a channel that is NOT the default channel of the package:


Screen Shot 2021-09-16 at 16 36 19
Screen Shot 2021-09-16 at 16 36 04
Screen Shot 2021-09-16 at 16 35 28

Comment (ankitathomas) [reply pkg/sqlite/load.go:1326]: I verified this, the inner join was missing a compare on package name.
`SELECT head_operatorbundle_name FROM channel
    INNER JOIN package ON channel.name = package.default_channel
    AND channel.package_name = package.name
    WHERE channel.head_operatorbundle_name = ?`

File: pkg/sqlite/query.go (hunk 1 above)
Comment (ankitathomas) [reply pkg/sqlite/query.go:1348]: Same thing here, the inner join was missing a compare on `channel.package_name = package.name`, this looks good.

**Review by ankitathomas — APPROVED** [review]
Thanks for catching this, the change looks good!

File: pkg/sqlite/load.go [hunk 3]
```diff
@@ -1320,9 +1320,10 @@ func (s *sqlLoader) rmChannelEntry(tx *sql.Tx, csvName string) error {
 
 func getTailFromBundle(tx *sql.Tx, head string) (bundles []string, err error) {
 	getReplacesSkips := `SELECT replaces, skips FROM operatorbundle WHERE name=?`
-	isDefaultChannelHead := `SELECT head_operatorbundle_name FROM channel
-							INNER JOIN package ON channel.name = package.default_channel
-							WHERE channel.head_operatorbundle_name = ?`
+	isDefaultChannelHead := `SELECT head_operatorbundle_name FROM package, channel
```
Comment (perdasilva) [inline pkg/sqlite/load.go:1323]: I took this issue as an opportunity to dust off my SQL sk1llz and here's what I've found:

The issue in the original query seems to be that the inner join is only made against the channel name and there are two channels (2.x and alpha). So, it's also joining the alpha rows with packages that have a default channel of alpha. An alternative query here would be to join on both the package name and the default channel. E.g.

```
SELECT head_operatorbundle_name FROM channel
INNER JOIN package ON channel.package_name = package.name AND channel.name = package.default_channel
WHERE channel.head_operatorbundle_name = ?
```

I believe the above and what you propose are equivalent in performance and intent. However, [this](https://stackoverflow.com/questions/13476029/multiple-table-select-vs-join-performance) StackOverflow entry seems to suggest that the inner join syntax is recommended. Not that this means you should change anything. Just relating what I've found ^^

File: bundles/apicurio-registry-operator.v1.0.0-v2.0.0.final/manifests/apicurio-registry-operator_v1_serviceaccount.yaml [hunk 4]
```diff
@@ -0,0 +1,9 @@
+apiVersion: v1
```
Comment (perdasilva) [inline bundles/apicurio-registry-operator.v1.0.0-v2.0.0.final/manifests/apicurio-registry-operator_v1_serviceaccount.yaml:1]: Just a question to help my understanding: why was it necessary to add all of these manifest and bundle files? why were they not there in the first place?

File: pkg/sqlite/load.go (hunk 3 above)
Comment (perdasilva) [reply pkg/sqlite/load.go:1323]: Oh, I didn't see the previous discussion on this. Feel free to ignore the comments above ^^

File: bundles/apicurio-registry-operator.v1.0.0-v2.0.0.final/manifests/apicurio-registry-operator_v1_serviceaccount.yaml (hunk 4 above)
Comment (camilamacedo86) [reply bundles/apicurio-registry-operator.v1.0.0-v2.0.0.final/manifests/apicurio-registry-operator_v1_serviceaccount.yaml:1]: I added the scenario for we are able to add a new test to ensure that after the change we not only break anything as it is working as supposed to be. 

>

Comment (njhale) [reply bundles/apicurio-registry-operator.v1.0.0-v2.0.0.final/manifests/apicurio-registry-operator_v1_serviceaccount.yaml:1]: Most of these files don't directly impact the test at all.

@camilamacedo86 could you minimize the set of files added in this PR. This should only require CSVs and annotation files, right?

**Review by njhale — CHANGES_REQUESTED** [review]

**Comment by kevinrizza** [comment]
This change lgtm minus Nick's comments around testing

**Comment by camilamacedo86** [comment]
HI @njhale and @kevinrizza,

That the manifests were removed from the mocks. Feel free to check.

**Pull request merged**